### PR TITLE
Feature: DCMAW 8381

### PR DESCRIPTION
**Pull request by ThomasIent**
# DCMAW-8381: GA4 schema on the Sign in page

- [Ticket DCMAW-8381](https://govukverify.atlassian.net/browse/DCMAW-8381)
- [Logging project PR 1](https://github.com/govuk-one-login/mobile-android-logging/pull/95)
- [Logging project PR 2](https://github.com/govuk-one-login/mobile-android-logging/pull/99) 
- [Logging project PR 3](https://github.com/govuk-one-login/mobile-android-logging/pull/104) 
- [GA4 Schema v3.1](https://govukverify.atlassian.net/wiki/spaces/PI/pages/3790995627/GA4+One+Login+Mobile+Application+Data+Schema+V3.1)

## Description of changes:

- Adds Crashlytics
- Integrates the logging project.

### Crashlytics

DCMAW-8381-crashlytics

### Analytics

AC1a: trackScreen analytics [opt-in]
GIVEN I have previously opted-in to analytics
AND I’m on the Sign in page
THEN the trackScreen analytics defined in the OpenAPI spec will be sent to GA4
AND the firebaseScreenId is set to 30a6b339-75a8-44a2-a79a-e108546419bf

DCMAW-8381-ScreenView-1
DCMAW-8381-ScreenView-2

AC2a: trackEventLink analytics (Sign in button) [opt-in]
GIVEN I have previously opted-in to analytics
AND I’m on the Sign in page
WHEN I tap the Sign in CTA
THEN the trackEventLink analytics defined in the OpenAPI spec will be sent to GA4
AND my user properties contain my device, brand, model and operating system - this is important for tracking across the WebView!

DCMAW-8381-Navigation-1
DCMAW-8381-Navigation-2

AC1b: trackScreen analytics [opt-out]
GIVEN I have not previously opted-in to analytics
AND I’m on the Sign in page
THEN no data is sent to GA4

DCMAW-8381-Disabled-2

AC2b: trackEventLink analytics (Sign in button) [opt-out]
GIVEN I have not previously opted-in to analytics
AND I’m on the Sign in page
WHEN I tap the Sign in CTA
THEN no data is sent to GA4

DCMAW-8381-Disabled

AC4: Welsh language analytics
GIVEN user is on the Sign in page
WHEN the language parameter is set to CY
THEN all the analytics is still sent in English

DCMAW-8381-Welsh-Navigation
DCMAW-8381-Welsh-ScreenView

### Before creating the pull request

- [x] Commit messages that conform to conventional commit messages.
- [x] Ran the app locally ensuring it builds.
- [x] Tests pass locally.
- [x] Pull request has a clear title with a short description about the feature or update.
- [x] Created a `draft` pull request if it's not ready for review.

### Before the CODEOWNERS review the pull request

- [x] Complete all Acceptance Criteria within Jira ticket.
- [x] Self-review code.
- [x] Successfully run changes on a testing device.
- [x] Complete automated Testing:
  * [x] Unit Tests.
  * [x] Integration Tests.
  * [x] Instrumentation / Emulator Tests.
- [x] Review [Accessibility considerations].
- [x] Handle PR comments.

### Before merging the pull request

- [x] [Sonar cloud report] passes inspections for your PR.
- [x] Resolve all comments.

[Sonar cloud report]: https://sonarcloud.io/project/overview?id=di-mobile-android-onelogin-app
[Accessibility considerations]: https://developer.android.com/guide/topics/ui/accessibility/testing

### CODE REVIEW

**Comment by ThomasIent** [comment]
Evidence has now been updated for the English and Welsh ACs, I left the disabled as it doesn't matter what the parameters were.

File: app/src/main/java/uk/gov/onelogin/analytics/data/adpater/AndroidAnalyticsAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package uk.gov.onelogin.analytics.data.adpater
+
+import android.util.Log
+import uk.gov.onelogin.analytics.domain.logger.AnalyticsLogger
+import uk.gov.onelogin.analytics.domain.model.AnalyticsEvent
+
+fun interface LogWriter {
```
[the remark this answers is not in the document]
Comment (ThomasIent) [reply app/src/main/java/uk/gov/onelogin/analytics/data/adpater/AndroidAnalyticsAdapter.kt:7]: Not if we need it to implement the local types, no. If I was able to use the stuff in the logging project then yes, I would have used it. It is based on that stuff, for that reason.

File: app/src/main/java/uk/gov/onelogin/analytics/data/adpater/AndroidAnalyticsAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        event.parameters.forEach { (t, u) ->
+            builder.append(String.format(PARAM, t, u))
+        }
+        builder.append(PARAMS_END)
+        log.write(TAG, builder.toString())
+    }
+
+    companion object {
+        const val TAG = "Analytics Logged"
+        const val EVENT_TITLE = "EventType: %s "
```
[the remark this answers is not in the document]
Comment (ThomasIent) [reply app/src/main/java/uk/gov/onelogin/analytics/data/adpater/AndroidAnalyticsAdapter.kt:28]: Sure. It gets tidied up a bit in 9613, but could do with even more tidying tbh.

File: app/src/main/java/uk/gov/onelogin/analytics/data/adpater/FirebaseAnalyticsAdapter.kt
[the remark this answers is not in the document]
Comment (ThomasIent) [reply app/src/main/java/uk/gov/onelogin/analytics/data/adpater/FirebaseAnalyticsAdapter.kt:1]: The FirebaseAnalyticsLogger uses and expects crashlytics, using it crashes the app. I was working with and around both classes until I realised I wasn't calling any of their methods and just cut them completely.

File: app/src/main/java/uk/gov/onelogin/analytics/data/adpater/AndroidAnalyticsAdapter.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ThomasIent) [reply app/src/main/java/uk/gov/onelogin/analytics/data/adpater/AndroidAnalyticsAdapter.kt:7]: This whole class is just an example of how we can add extra AnalyticsLogger.Adapters it isn't actually necessary as we can get all the Firebase logs from using adb commands, which is why it is a little sloppy with the actual output

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -46,7 +46,13 @@
                 <data android:path="/redirect"/>
             </intent-filter>
         </activity>
+
         <meta-data android:name="firebase_analytics_collection_enabled" android:value="false" />
+        <meta-data android:name="firebase_crashlytics_collection_enabled" android:value="false" />
```
Comment (alex-bradbury-gds) [inline app/src/main/AndroidManifest.xml:51]: do we want this to be turned off?

Comment (obinns-dd) [reply app/src/main/AndroidManifest.xml:51]: Yes, this prevents analytics / crashes being sent before the user has opted in.

Comment (ThomasIent) [reply app/src/main/AndroidManifest.xml:51]: It matches what id-check are doing and where the `enable` method on the FirebaseAnalyticsLogger calls both FirebaseAnalytics and FirebaseCrashlytics, ensures that they are synchronised at all times. I don't think there is any legal / GDPR reason for doing this with crashlytics (unlike the Analytics), more it is how the code expects it to be done.

**Review by alex-bradbury-gds — APPROVED** [review]
Looks really good!

**Comment by co-eoin** [comment]
QA Reviewed & Approved ✅

**Pull request merged**